--- a/run_rustc/Makefile
+++ b/run_rustc/Makefile
@@ -82,9 +82,9 @@
 endif

 ../output$(OUTDIR_SUF)/rustc:
-	$(MAKE) -C ../ output$(OUTDIR_SUF)/rustc -j 3
+	$(MAKE) -C ../ output$(OUTDIR_SUF)/rustc -j $(NIX_BUILD_CORES)
 ../output$(OUTDIR_SUF)/cargo:
-	$(MAKE) -C ../ output$(OUTDIR_SUF)/cargo -j 3
+	$(MAKE) -C ../ output$(OUTDIR_SUF)/cargo -j $(NIX_BUILD_CORES)

 $(BINDIR)rustc_m: ../output$(OUTDIR_SUF)/rustc
 	@mkdir -p $(dir $@)
@@ -125,10 +125,10 @@
 	@mkdir -p $(LIBDIR)
 ifneq ($(RUSTC_VERSION),1.39.0)
 	@echo [CARGO] $(RUST_SRC)libstd/Cargo.toml '>' $(OUTDIR)build-std
-	$VCARGO_TARGET_DIR=$(OUTDIR)build-std RUSTC=$(BINDIR_2)rustc $(CARGO_ENV) $(BINDIR)cargo build --manifest-path $(RUST_SRC)libstd/Cargo.toml  -j 1 --release --features panic-unwind -v
+	$VCARGO_TARGET_DIR=$(OUTDIR)build-std RUSTC=$(BINDIR_2)rustc $(CARGO_ENV) $(BINDIR)cargo build --manifest-path $(RUST_SRC)libstd/Cargo.toml  -j $(NIX_BUILD_CORES) --release --features panic-unwind -v
 endif
 	@echo [CARGO] $(RUST_SRC)libtest/Cargo.toml '>' $(OUTDIR)build-std
-	$VCARGO_TARGET_DIR=$(OUTDIR)build-std RUSTC=$(BINDIR_2)rustc $(CARGO_ENV) $(BINDIR)cargo build --manifest-path $(RUST_SRC)libtest/Cargo.toml -j 1 --release --features panic-unwind -v
+	$VCARGO_TARGET_DIR=$(OUTDIR)build-std RUSTC=$(BINDIR_2)rustc $(CARGO_ENV) $(BINDIR)cargo build --manifest-path $(RUST_SRC)libtest/Cargo.toml -j $(NIX_BUILD_CORES) --release --features panic-unwind -v
 	$Vcp --remove-destination $(OUTDIR)build-std/release/deps/*.rlib $(LIBDIR)
 	$Vcp --remove-destination $(OUTDIR)build-std/release/deps/*.so $(LIBDIR)

@@ -136,10 +136,10 @@
 $(BINDIR)rustc: $(BINDIR)rustc_m $(BINDIR)cargo $(CARGO_HOME)config $(LIBDIR_S)libstd.rlib
 	@mkdir -p $(PREFIX)tmp
 	@echo [CARGO] $(RUST_SRC)rustc/Cargo.toml '>' $(OUTDIR)build-rustc
-	$V$(RUSTC_ENV_VARS) TMPDIR=$(abspath $(PREFIX)tmp) CARGO_TARGET_DIR=$(OUTDIR)build-rustc RUSTC=$(BINDIR_S)rustc RUSTC_ERROR_METADATA_DST=$(abspath $(PREFIX)) $(CARGO_ENV) $(BINDIR)cargo build --manifest-path $(RUST_SRC)rustc/Cargo.toml --release -j 1
+	$V$(RUSTC_ENV_VARS) TMPDIR=$(abspath $(PREFIX)tmp) CARGO_TARGET_DIR=$(OUTDIR)build-rustc RUSTC=$(BINDIR_S)rustc RUSTC_ERROR_METADATA_DST=$(abspath $(PREFIX)) $(CARGO_ENV) $(BINDIR)cargo build --manifest-path $(RUST_SRC)rustc/Cargo.toml --release -j $(NIX_BUILD_CORES)
 ifeq ($(RUSTC_VERSION),1.39.0)
 	@echo [CARGO] $(RUST_SRC)librustc_codegen_llvm/Cargo.toml '>' $(OUTDIR)build-rustc
-	$V$(RUSTC_ENV_VARS) TMPDIR=$(abspath $(PREFIX)tmp) CARGO_TARGET_DIR=$(OUTDIR)build-rustc RUSTC=$(BINDIR_S)rustc RUSTC_ERROR_METADATA_DST=$(abspath $(PREFIX)) $(CARGO_ENV) $(BINDIR)cargo rustc --manifest-path $(RUST_SRC)librustc_codegen_llvm/Cargo.toml --release -j 1 -- -L $(abspath $(OUTDIR)build-rustc/release/deps)
+	$V$(RUSTC_ENV_VARS) TMPDIR=$(abspath $(PREFIX)tmp) CARGO_TARGET_DIR=$(OUTDIR)build-rustc RUSTC=$(BINDIR_S)rustc RUSTC_ERROR_METADATA_DST=$(abspath $(PREFIX)) $(CARGO_ENV) $(BINDIR)cargo rustc --manifest-path $(RUST_SRC)librustc_codegen_llvm/Cargo.toml --release -j $(NIX_BUILD_CORES) -- -L $(abspath $(OUTDIR)build-rustc/release/deps)
 	@mkdir -p $(LIBDIR)../codegen-backends
 	@echo "[CP] libraries and results"
 	$V$(call fn_opt_copy,$(OUTDIR)build-rustc/release/librustc_codegen_llvm.so,$(LIBDIR)../codegen-backends/librustc_codegen_llvm-llvm.so)
